MetaMask/metamask-extension
# Fix autolock field to accept decimals in Firefox

## Explanation



The autolock field on the Settings screen — the field that allows users to set the duration that MetaMask will wait for until automatically locking — does not always accept decimal numbers. This breaks the e2e test for this feature as it attempts to set this field to "0.1".

More specifically, the React component responsible for this field passes whatever the user inputs through the `Number` function immediately and then uses this to repopulate the input. Therefore, if the user enters "3" followed by a ".", `Number("3.")` will be called. This evaluates to the number 3, and "3" becomes the new value of the field. As a result, the "." can never be typed.

Curiously, this behavior only happens in Firefox; Chrome seems to keep the "." in the input field when it's typed. This happens because `onChange` event doesn't seem to get fired until a number is typed *after* the ".". This may be due to underlying differences in the DOM between Chrome and Firefox.

Regardless, always passing the input through `Number` creates other odd behavior, such as the fact that the input can never be cleared (because `Number("")` evaluates to 0).

This commit solves these problems by saving the "raw" version of the user's input as well as the normalized version. The raw version is always used to populate the input, whereas the normalized version is saved in state.

## Screenshots/Screencaps



### Before



#### Chrome

It's hard to see, but 1) pressing Backspace when it says "0" does nothing and 2) adding numbers after the decimal does nothing.

https://github.com/MetaMask/metamask-extension/assets/7371/6612b8b0-13e7-41dc-937c-7c6aeff7dc65

#### Firefox

It's hard to see, but 1) pressing Backspace when it says "0" does nothing, 2) adding decimals or anything after the decimal resets the field to "0".

https://github.com/MetaMask/metamask-extension/assets/7371/4dd3a5fe-5557-4570-89f9-057df9d6d2cf

### After



#### Chrome

https://github.com/MetaMask/metamask-extension/assets/7371/c9ac9b7c-5521-465e-87ee-fa4947dd4ef7

#### Firefox

https://github.com/MetaMask/metamask-extension/assets/7371/11292675-939f-4c3d-b9ce-c01635371eda

## Manual Testing Steps



Fundamentally you should ensure that if you go into Settings > Advanced, update the autolock setting to 1, save, then wait 1 minute, MetaMask should automatically lock. You should also see an error (and should not be able to save) if you enter a number past 10080.

With this change, however, you should also be able to enter 0.1, and MetaMask should autolock in 5 seconds.

Also, if you empty the field, it should default to 5 minutes (this has always been the placeholder value for this input, it's just that it never got used).

All of this behavior should be the same in both Chrome and Firefox.

Past this, all other settings on the page should work as intended.

## Pre-merge author checklist


## Pre-merge reviewer checklist


If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.
